drivers/SmartThings/harman-luxury/src/lustre/ws.lua

```diff
@@ -0,0 +1,655 @@
+    log.error(self.id, "Invalid client websocket")
+    return nil, "only a client can connect"
+  end
+  if not host or not port then
+    return nil, "missing host or port"
+  end
+  log.debug(self.id, "calling socket.connect")
+  local r, err = self.socket:connect(host, port)
+  log.debug(self.id, "Socket connect completed",
+    r or err)
+  if not r then
+    return nil, "socket connect failure: " .. err
+  end
+
+  return self:client_handshake_and_start(host, port)
+end
+
+function WebSocket:accept()
+  return nil, "Not yet implemented"
+end
+
+---@param close_code CloseCode
+---@param reason string
+---@return number 1 if success
+---@return string|nil
+function WebSocket:close(close_code, reason)
+  log.debug("sending close message",
+    close_code.type or close_code.value, reason)
+  if self.state == "Active" then
+    local close_frame = Frame.close(close_code,
+      reason):set_mask() -- TODO client vs server
+    local tx, reply = cosock.channel.new()
+    reply:settimeout(0.5)
+    log.debug("sending frame to socket task")
+    local suc, err = self._send_tx:send(
+      {frame = close_frame, reply = tx})
+    log.debug("sent frame to socket task", suc,
+      err)
+    if not suc then
+      return nil, "channel error:" .. err
+    end
+    log.debug("waiting on reply")
+    local reply = reply:receive()
+    log.debug("reply received")
+    return reply
+  elseif self.state == "ClosedBySelf" then
+    self.state = "CloseAcknowledged"
+  end
+
+  return 1, log.debug("closed websocket")
+end
+
+---Cosock internal interface for using `cosock.socket.select`
+---@param kind string
+---@param waker fun()
+function WebSocket:setwaker(kind, waker)
+  assert(kind == "recvr",
+    "unsupported wake kind: " .. tostring(kind))
+  assert(self._waker == nil or waker == nil,
+    "waker already set, receive can't be waited on from multiple places at once")
+  self._waker = waker
+
+  -- if messages waiting, immediately wake
+  if #self._recv_tx.link.queue > 0 and waker then
+    waker()
+  end
+end
+
+---Spawn the receive loop
+---@return string|nil
+function WebSocket:_receive_loop()
+  log.trace(self.id, "starting receive loop")
+  local loop_state = {
+    partial_frames = nil,
+    received_bytes = 0,
+    frames_since_last_ping = 0,
+    pending_pongs = 0,
+    multiframe_message = false,
+    utf8_check_backward_idx = 0,
+    msg_type = nil,
+  }
+  local order = false
+  while self.state ~= "CloseAcknowledged"
+    and self.state ~= "Terminated" do
+    log.trace(self.id, "loop top")
+    local rs = (order
+                 and {self._send_rx, self.socket})
+                 or {self.socket, self._send_rx}
+    order = not order
+    local recv, _, err = socket.select(rs, nil,
+      self.config._keep_alive)
+    log.debug((recv and "recv") or "~recv",
+      err or "")
+    if not recv then
+      if self:_handle_select_err(loop_state, err) then
+        return
+      end
+    elseif self:_handle_recvs(loop_state, recv, 1) then
```

Comment:
I'm the person that was helping you out on the forums :)

You can fix this issue by only including this file while preserving the directory structure. You don't need to include the entire library.

Reply:
Okay. I'll modify  it